Plugins/Data/NWScript/inc_array.nss

```diff
@@ -0,0 +1,485 @@
+
+/// @defgroup data_array_set Array Set
+/// @brief Sets the element at the index, where size > index >= 0.
+/// @ingroup data
+/// @param obj The object.
+/// @param tag The tag.
+/// @param index The index.
+/// @param element The element.
+/// @{
+void Array_Set_Flt(string tag, int index, float  element, object obj=OBJECT_INVALID);
+void Array_Set_Int(string tag, int index, int    element, object obj=OBJECT_INVALID);
+void Array_Set_Obj(string tag, int index, object element, object obj=OBJECT_INVALID);
+void Array_Set_Str(string tag, int index, string element, object obj=OBJECT_INVALID);
+/// @}
+
+/// @}
+
+//
+// Local Utility Functions.
+//
+string GetTableName(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    return "array_" + ObjectToString(obj) + "_" + tag;
+}
+
+string GetTableCreateString(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    // for simplicy sake, everything is turned into a string.  Possible enhancement
+    // to create specific tables for int/float/whatever.
+    return "CREATE TABLE IF NOT EXISTS " + GetTableName(tag, obj) + " ( ind INTEGER PRIMARY KEY, value TEXT)";
+}
+
+int TableExists(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    string stmt = "SELECT name FROM sqlite_master WHERE type = 'table' AND name = '" + GetTableName(tag, obj) + "'";
+    sqlquery sqlQuery = SqlPrepareQueryObject(GetModule(), stmt);
+    return SqlStep(sqlQuery);
+}
+
+void ExecuteStatement(string statement, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    // There's no direct "execute this.."  everything has to be prepared then executed.
+    //WriteTimestampedLogEntry("SQL: " + statement);
+    sqlquery sqlQuery = SqlPrepareQueryObject(GetModule(), statement);
+    SqlStep(sqlQuery);
+}
+
+void CreateArrayTable(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    string createStatement = GetTableCreateString(tag, obj);
+    ExecuteStatement(createStatement, obj);
+}
+
+// Get the table row count.  Returns -1 on error (0 is a valid number of rows in a table)
+int GetRowCount(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    CreateArrayTable(tag, obj);
+    string stmt = "SELECT COUNT(1) FROM " + GetTableName(tag, obj);
+    sqlquery sqlQuery = SqlPrepareQueryObject(GetModule(), stmt);
+    if ( SqlStep(sqlQuery) ) {
+        return SqlGetInt(sqlQuery, 0);
+    }
+    return -1;
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+// return the value contained in location "index"
+string Array_At_Str(string tag, int index, object obj=OBJECT_INVALID)
+{
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    if (!TableExists(tag, obj)) {
```

Comment:
Why do we do the init here? Can't you just attempt a query, and if it doesn't exist it'll return `""`.

Reply:
restructured it a bit..  basically just calls "create table if not exists" a lot...   comments welcome